### PR TITLE
Move album art when renaming [#237] 

### PR DESCRIPTION
**Pull request by elbeardmorez**
Hi,

**Here I aim to improve the file renaming experience for users with external album art**

I've attempted to move all 'relevant' album art from a music file's source directory to its target directory when it is renamed. Which art is moved is dependent on the '[albumart] -> filenames' config option which allows for wildcards and quodlibet constants, e.g. **front.jpg,back.jpg,cd\*.jpg,\<artist\>.jpg** via a fairly basic priority first matters algorithm. Although no doubt there will be situations / setups that this change won't fully support / handle, my thought is certainly 'anything is better than nothing', and aside from the algorithm not looking in subfolders, I struggle to see much that the flexibility of priority/wildcards/\<constants\> couldn't handle.

commit: move album art when renaming [#237]

image: [additional rename options](http://picpaste.com/00e9db5a6b0a9d35d9e8e30acb4d1206.png)
_**note:**_ for this patch, in attempt to find corner cases I tested:
-with/without custom config entries
-rel/abs target paths
-target with/without audio
-target with/without images
-multiple art files in source
-multiple source folders with art
_**note2:**_ for consideration:
-use of shutil (uses 'copy2'). safer but issues with file attribute changes
-rather trivial attempt to ease scaling with 'art sets' dict
-no testing in either python3 or non gnu/linux environments at all sorry!
_**note3:**_ limitations:
-no support for image names specified as <b>both</b> template and wildcard .e.g. \<artist\>*.jpg

**With** the changes, I **don't** have to manually navigate to folders and construct a move command for the album art files ..which is very painful for me! ..**without**, I **do** have to. That is why I wish this to be included in quodlibet.

**Further implementation notes:**
-Everywhere possible I've left defaults as they were and disabled the additional features.

**Configs**
-My current config mods, not required in any way, but as an example..

[albumart]
filenames = front.jpg,back.jpg,cd*.jpg,\<artist\>.jpg

[rename]
moveart = true
moveart_overwrite = true
removeemptydirs = true

### CODE REVIEW

File: quodlibet/quodlibet/config.py [hunk 1]
```diff
@@ -215,6 +215,9 @@ def _config_text(text):
         "spaces": "false",
         "windows": "true",
         "ascii": "false",
+        "moveart": "false",
```
Comment (declension) [inline quodlibet/quodlibet/config.py]: `move_art` and `move_art_overwrite` seem better IMO

File: quodlibet/quodlibet/qltk/renamefiles.py [hunk 2]
```diff
@@ -282,6 +336,77 @@ def __rename(self, library):
         library.changed(was_changed)
         self.save.set_sensitive(False)
 
+    def __moveart(self, art_sets, pathfile_old, pathfile_new, song):
+
+        path_old = os.path.dirname(os.path.realpath(pathfile_old))
+        path_new = os.path.dirname(os.path.realpath(pathfile_new))
+        if os.path.samefile(path_old, path_new):
+            return
+        if (path_old in art_sets.keys() and len(art_sets[path_old]) == 0):
```
Comment (declension) [inline quodlibet/quodlibet/qltk/renamefiles.py]: would `and not art_sets[path_old]` suffice?

File: quodlibet/quodlibet/qltk/renamefiles.py [hunk 3]
```diff
@@ -282,6 +336,77 @@ def __rename(self, library):
...
+            return
+
+        # get art set for path
+        images = []
+        if path_old in art_sets.keys():
+            images = art_sets[path_old]
+        else:
+            def glob_escape(s):
+                for c in ['[', '*', '?']:
```
Comment (declension) [inline quodlibet/quodlibet/qltk/renamefiles.py]: BTW neater is probably `for c in '[*?':`

File: quodlibet/quodlibet/qltk/renamefiles.py [hunk 4]
```diff
@@ -282,6 +336,77 @@ def __rename(self, library):
...
+                    s = s.replace(c, '[' + c + ']')
+                return s
+
+            # generate art set for path
+            art_sets[path_old] = images
+            path_old_escaped = glob_escape(path_old)
+            for suffix in self.IMAGE_EXTENSIONS:
+                images.extend(glob.glob(os.path.join(path_old_escaped,
+                                                     "*." + suffix)))
+        if len(images) > 0:
```
Comment (declension) [inline quodlibet/quodlibet/qltk/renamefiles.py]: `if images` nicer IMO

File: quodlibet/quodlibet/config.py [hunk 5]
```diff
@@ -257,6 +260,7 @@ def _config_text(text):
         "prefer_embedded": "false",
         "force_filename": "false",
         "filename": "folder.jpg",
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/config.py]: can't quite tell from the snippet, but if you were thinking _'filename' -> 'filenames'_, then they are actually two very separate config entries. the first is for the chosen filename for album art show in the playbar (thus unrelated to this patch set), the latter is the new, additional config entry denoting 'search filenames'. if i'm off the mark here please shout! i'll change _'filenames' -> 'search_filenames'_ to make this clearer regardless (reason i didn't do this before is that i have to change it elsewhere too (e.g. the coverswidgetbar work))

File: quodlibet/quodlibet/config.py [hunk 6]
```diff
@@ -215,6 +215,9 @@ def _config_text(text):
         "spaces": "false",
         "windows": "true",
         "ascii": "false",
+        "moveart": "false",
+        "moveart_overwrite": "false",
+        "removeemptydirs": "false",
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/config.py]: done

File: quodlibet/quodlibet/config.py (hunk 1 above)
Comment (elbeardmorez) [reply quodlibet/quodlibet/config.py]: done

File: quodlibet/quodlibet/qltk/renamefiles.py [hunk 7]
```diff
@@ -236,13 +269,21 @@ def __rename(self, library):
         was_changed = set()
         skip_all = False
         self.view.freeze_child_notify()
+        moveart = config.getboolean("rename", "moveart")
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/renamefiles.py]: yes, certainly would help clarity

File: quodlibet/quodlibet/qltk/renamefiles.py (hunk 3 above)
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/renamefiles.py]: hehe, er, yes.. :)

File: quodlibet/quodlibet/qltk/renamefiles.py [hunk 8]
```diff
@@ -236,13 +269,21 @@ def __rename(self, library):
...
+            if entry.new_name is None:
+                continue
             song = entry.song
-            new_name = entry.new_name
             old_name = entry.name
-            if new_name is None:
-                continue
+            old_pathfile = song['~filename']
+            new_name = entry.new_name
+            new_pathfile = new_name if (
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/renamefiles.py]: yep, simplified and added comments

File: quodlibet/quodlibet/qltk/renamefiles.py (hunk 2 above)
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/renamefiles.py]: yep, first chunk of Python for a *very* long time! the latter stuff shouldn't be this poor (hopefully anyway!)

File: quodlibet/quodlibet/qltk/renamefiles.py (hunk 4 above)
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/renamefiles.py]: yep sorry, again!

**Comment by elbeardmorez** [comment]
layout updated -> [move art options](http://picpaste.com/1e1ec65feccc0e3d76aad822fdfd1e42.png)

**Comment by declension** [comment]
Thanks for the hard work!

File: quodlibet/quodlibet/qltk/renamefiles.py [hunk 9]
```diff
@@ -282,6 +317,70 @@ def __rename(self, library):
...
+        # get art set for path
+        images = []
+        if path_old in art_sets.keys():
+            images = art_sets[path_old]
+        else:
+            art_sets[path_old] = images
+            # generate art set for path
+            for suffix in self.IMAGE_EXTENSIONS:
+                images.extend(glob.glob(os.path.join(path_old, "*." + suffix)))
+        if len(images) > 0:
```
Comment (elbeardmorez) [inline quodlibet/quodlibet/qltk/renamefiles.py]: @declension, see below, `if '<' in ..`

**Pull request merged**